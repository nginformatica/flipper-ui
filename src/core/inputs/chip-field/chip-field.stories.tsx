import React, { useState } from 'react'
import type { Meta, StoryFn } from '@storybook/react'
import { ChipField } from '.'

export default {
    title: 'Inputs/ChipField',
    component: ChipField
} as Meta<typeof ChipField>

const Template: StoryFn<typeof ChipField> = args => {
    const [state, setState] = useState([{ value: 'alpha' }])

    const handleRemove = (value: string) =>
        setState(state.filter(item => item.value !== value))

    const handleAdd = (item: string) => setState([{ value: item }, ...state])

    return (
        <ChipField
            {...args}
            values={state}
            onDelete={handleRemove}
            onAdd={handleAdd}
        />
    )
}

export const Default = Template.bind({})

Default.args = {
    variant: 'outlined',
    label: 'chips'
}
