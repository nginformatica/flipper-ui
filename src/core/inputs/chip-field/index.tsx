import React, { useMemo } from 'react'
import type { ReactNode } from 'react'
import ChipInput from 'material-ui-chip-input'
import { styled } from 'styled-components'
import type { InputLabelProps } from '@material-ui/core'
import type { ChipRenderer } from 'material-ui-chip-input'

export interface ChipFieldProps<T extends TChipValues> {
    allowDuplicates?: boolean
    values: T[]
    variant?: 'outlined' | 'standard' | 'filled'
    disabled?: boolean
    fullWidth?: boolean
    fullWidthInput?: boolean
    InputLabelProps?: InputLabelProps
    label?: ReactNode
    placeholder?: string
    readOnly?: boolean
    error?: boolean
    chipRenderer?: ChipRenderer
    onAdd: (values: string) => void
    onDelete: (value: string, index?: number) => void
}

export type TChipValues = {
    value: string | number
}

const StyledChipInput = styled(ChipInput)`
    & > div > div {
        padding: 0px !important;
    }
    & > div > div > div {
        margin: 2px;
    }
    & > label {
        top: -6px !important;
        font-size: 14px !important;
    }
    & > label.MuiInputLabel-shrink {
        top: 0px !important;
    }
    & > div > div > input {
        padding: 10px !important;
        font-size: 14px !important;
    }
`

export const ChipField = <T extends TChipValues>(props: ChipFieldProps<T>) => {
    const values = useMemo(
        () => props.values.map(item => item.value),
        [props.values]
    )

    return (
        <StyledChipInput
            {...props}
            alwaysShowPlaceholder
            id='chip-field-root'
            value={values}
        />
    )
}

export default ChipField
