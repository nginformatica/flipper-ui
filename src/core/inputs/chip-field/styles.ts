import ChipInput from 'material-ui-chip-input'
import { default as styled } from 'styled-components'

export const StyledChipInput = styled(ChipInput)`
    & > div > div {
        padding: 0px !important;
    }
    & > div > div > div {
        margin: 2px;
    }
    & > label {
        top: -6px !important;
        font-size: 14px !important;
    }
    & > label.MuiInputLabel-shrink {
        top: 0px !important;
    }
    & > div > div > input {
        padding: 10px !important;
        font-size: 14px !important;
    }
`
