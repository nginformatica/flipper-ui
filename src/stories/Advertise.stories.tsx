import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import Advertise from '../core/Advertise'

export default {
    title: 'Advertise',
    component: Advertise
} as ComponentMeta<typeof Advertise>

const Template: ComponentStory<typeof Advertise> = args => (
    <Advertise { ...args } />
)

export const Default = Template.bind({})

Default.args = {
    comment:
        'Be like water. You put water into a cup. \
        It becomes the cup. You put water into a bottle. It becomes the bottle.',
    author: 'Bruce Lee',
    children: 'Advertise'
}
