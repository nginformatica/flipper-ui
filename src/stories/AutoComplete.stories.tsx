import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import AutoComplete from '../core/AutoComplete'
import ListItem from '../core/ListItem'
import TextField from '../core/TextField'
import Typography from '../core/Typography'

export default {
    title: 'AutoComplete',
    component: AutoComplete
} as ComponentMeta<typeof AutoComplete>

const Template: ComponentStory<typeof AutoComplete> = args => (
    <AutoComplete {...args} />
)

export const Default = Template.bind({})

Default.args = {
    suggestions: [
        { value: 'rock', label: 'Rock' },
        { value: 'pop', label: 'Pop' },
        { value: 'classic', label: 'Classic' }
    ],
    renderInput: props => (
        <TextField fullWidth placeholder='Music genres' {...props} />
    ),
    renderSuggestion: (item, props, selected) => (
        <ListItem key={item} selected={selected} {...props}>
            {item.label}
        </ListItem>
    ),
    onChange: () => null
}

export const OpeningWhenFocused = () => (
    <AutoComplete
        openOnFocus
        value={{ value: 'classic', label: 'Classic' }}
        suggestions={[
            { value: 'rock', label: 'Rock' },
            { value: 'pop', label: 'Pop' },
            { value: 'classic', label: 'Classic' }
        ]}
        renderInput={props => (
            <TextField fullWidth placeholder='Music genres' {...props} />
        )}
        renderSuggestion={(item, props, selected) => (
            <ListItem selected={selected} {...props}>
                {item.label}
            </ListItem>
        )}
        onChange={() => null}
    />
)

export const WithSubheader = () => (
    <AutoComplete
        openOnFocus
        suggestions={[
            { label: 'Genres', subheader: true },
            { value: 'rock', label: 'Rock' },
            { value: 'pop', label: 'Pop' },
            { value: 'classic', label: 'Classic' }
        ]}
        renderInput={props => (
            <TextField fullWidth placeholder='Music genres' {...props} />
        )}
        renderSuggestion={(item, props, selected) =>
            item.subheader ? (
                <ListItem>
                    <Typography variant='caption'>{item.label}</Typography>
                </ListItem>
            ) : (
                <ListItem {...props} selected={selected}>
                    {item.label || item}
                </ListItem>
            )
        }
        onChange={() => null}
    />
)
