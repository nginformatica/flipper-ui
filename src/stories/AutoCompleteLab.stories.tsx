import React from 'react'
import { ComponentMeta } from '@storybook/react'
import AutoComplete from '../core/AutoComplete'
import TextField from '../core/TextField'
import { Autocomplete } from '@material-ui/lab'

export default {
    title: 'AutoCompleteLab',
    component: AutoComplete
} as ComponentMeta<typeof AutoComplete>

export const Default = () => (
    <Autocomplete
        options={ [
            { label: 'Anemone', value: 'forget-me-not' },
            { label: 'Forget me not', value: 'anemone' },
            { label: 'Snapdragon', value: 'snapdragon' },
            { label: 'Sunflower', value: 'sunflower' },
            { label: 'Rose', value: 'rose' }
        ] }
        getOptionLabel={ option => option.label }
        style={ { width: 300 } }
        renderInput={ props => (
            <TextField
                { ...props }
                label='Flowers'
                placeholder='Flowers'
                variant='outlined'
                fullWidth
            />
        ) }
        onChange={ () => null }
    />
)
