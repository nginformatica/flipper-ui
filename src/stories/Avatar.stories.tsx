import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import Avatar from '../core/Avatar'

export default {
    title: 'Avatar',
    component: Avatar,
    argTypes: {
        backgroundColor: { control: 'color' },
        children: { control: 'text' }
    }
} as ComponentMeta<typeof Avatar>

const Template: ComponentStory<typeof Avatar> = args => <Avatar {...args} />

export const Default = Template.bind({})

export const Primary = Template.bind({})
Primary.args = {
    primary: true
}

export const WithChuildren = Template.bind({})
WithChuildren.args = {
    children: 'W',
    primary: true
}
