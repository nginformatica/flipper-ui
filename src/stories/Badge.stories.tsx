import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import Badge from '../core/Badge'
import Button from '../core/Button'

export default {
    title: 'Badge',
    component: Badge,
    argTypes: {
        backgroundColor: { control: 'color' },
        counter: { control: 'number' }
    }
} as ComponentMeta<typeof Badge>

const Template: ComponentStory<typeof Badge> = args => <Badge { ...args } />

export const Default = Template.bind({})
Default.args = {
    counter: 2,
    children: 'Badge'
}

export const Primary = Template.bind({})
Primary.args = {
    counter: 2,
    color: 'primary',
    children: 'Badge'
}

export const Secondary = Template.bind({})
Secondary.args = {
    counter: 150,
    color: 'secondary',
    children: (
        <Button variant='outlined'>
            Try changing the counter to Zero
        </Button>
    )
}

export const WithDot = Template.bind({})
WithDot.args = {
    counter: 1,
    variant: 'dot',
    color: 'primary',
    children: 'Badge'
}
