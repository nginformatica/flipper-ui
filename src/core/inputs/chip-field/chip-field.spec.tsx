import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { default as userEvent } from '@testing-library/user-event'
import '@testing-library/jest-dom'
import { MultipleField } from './chip-field.stories'

const mockOnChange = jest.fn()

describe('ChipField', () => {
    beforeEach(() => {
        mockOnChange.mockClear()
    })

    it('should render with the correct label', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i)

        expect(input).toBeInTheDocument()
    })

    it('should add a chip when Enter key is pressed', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i) as HTMLInputElement

        userEvent.type(input, 'New Chip{enter}').then(() =>
            fireEvent.keyDown(input, {
                key: 'Enter'
            })
        )

        waitFor(() => {
            expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object), [
                'New Chip'
            ])
        })

        waitFor(() => {
            expect(input.value).toBe('')
        })
    })

    it('should add a chip when comma is pressed', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i) as HTMLInputElement

        userEvent.type(input, 'New Chip').then(() =>
            fireEvent.keyDown(input, {
                key: ','
            })
        )

        waitFor(() => {
            expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object), [
                'New Chip'
            ])
        })

        waitFor(() => {
            expect(input.value).toBe('')
        })
    })

    it('should add a chip when space is pressed', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i) as HTMLInputElement

        userEvent.type(input, 'New Chip').then(() =>
            fireEvent.keyDown(input, {
                key: ' '
            })
        )

        waitFor(() => {
            expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object), [
                'New Chip'
            ])
        })

        waitFor(() => {
            expect(input.value).toBe('')
        })
    })

    it('should add a chip when semicolon is pressed', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i) as HTMLInputElement

        userEvent.type(input, 'New Chip').then(() =>
            fireEvent.keyDown(input, {
                key: ';'
            })
        )

        waitFor(() => {
            expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object), [
                'New Chip'
            ])
        })

        waitFor(() => {
            expect(input.value).toBe('')
        })
    })

    it('should not add a chip when a non-special key is pressed', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i) as HTMLInputElement

        userEvent.type(input, 'New Chip{a}').then(() =>
            fireEvent.keyDown(input, {
                key: 'a'
            })
        )

        waitFor(() => {
            expect(mockOnChange).not.toHaveBeenCalled()
        })

        waitFor(() => {
            expect(input.value).toBe('New Chipa')
        })
    })

    it('should not add a chip if input is empty', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i) as HTMLInputElement

        userEvent.type(input, '').then(() =>
            fireEvent.keyDown(input, {
                key: 'Enter'
            })
        )

        waitFor(() => {
            expect(mockOnChange).not.toHaveBeenCalled()
        })

        waitFor(() => {
            expect(input.value).toBe('')
        })
    })

    it('should call onChange with the correct values when multiple keys are pressed', () => {
        render(<MultipleField onChange={mockOnChange} />)

        const input = screen.getByLabelText(/Emails */i) as HTMLInputElement

        userEvent.type(input, 'Chip 1{comma}Chip 2{enter}')

        waitFor(() => {
            expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object), [
                'Chip 1',
                'Chip 2'
            ])
        })

        waitFor(() => {
            expect(input.value).toBe('')
        })
    })

    it('should match snapshot', () => {
        const { container } = render(<MultipleField onChange={mockOnChange} />)

        expect(container).toMatchSnapshot()
    })
})
