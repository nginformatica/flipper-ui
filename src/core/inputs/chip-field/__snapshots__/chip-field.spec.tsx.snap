// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipField should match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root sc-beySPh jTLxIV WAMuiChipInput-root-52"
    data-testid="chip-field-container"
    values="[object Object],[object Object],[object Object]"
  >
    <div
      class="WAMuiChipInput-standard-57 WAMuiChipInput-chipContainer-55 WAMuiChipInput-underline-65"
    >
      <div
        class="MuiButtonBase-root MuiChip-root WAMuiChipInput-chip-67 MuiChip-clickable MuiChip-deletable"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          alpha
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root WAMuiChipInput-chip-67 MuiChip-clickable MuiChip-deletable"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          beta
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiButtonBase-root MuiChip-root WAMuiChipInput-chip-67 MuiChip-clickable MuiChip-deletable"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          gamma
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiInputBase-root MuiInput-root WAMuiChipInput-inputRoot-53 WAMuiChipInput-standard-57 MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input WAMuiChipInput-input-54 WAMuiChipInput-standard-57"
          id="chip-field-root"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
